File: app/src/main/java/pl/orbitemobile/wspolnoty/activities/article/ArticlePresenter.java
/*
 * Copyright (c) 2017. All Rights Reserved. Michal Jankowski orbitemobile.pl
 */
package pl.orbitemobile.wspolnoty.activities.article;

import android.content.Context;
import android.content.Intent;
import android.view.MenuItem;

import io.reactivex.SingleObserver;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import pl.orbitemobile.wspolnoty.R;
import pl.orbitemobile.wspolnoty.activities.article.domain.ArticleScreen;
import pl.orbitemobile.wspolnoty.utilities.ConnectivityCheck;
import pl.orbitemobile.wspolnoty.utilities.DialogBuilder;
import pl.orbitemobile.wspolnoty.data.entities.Article;

public class ArticlePresenter implements ArticleContract.Presenter {
    
    public static final String ARTICLE_URL = "article_url";
    
    public static final String TITLE = "title";
    
    public static final String IMG_URL = "img_url";
    
    private final SingleObserver<Article> mArticleRemoteObserver = new ArticleRemoteObserver();
    
    private ArticleContract.View mView;
    
    private Context mContext;
    
    private ArticleScreen mUseCase;
    
    private CompositeDisposable mDisposable = new CompositeDisposable();
    
    private String mArticleUrl;
    
    private ConnectivityCheck mConnectivityCheck;
    
    public ArticlePresenter(final ArticleContract.View view, final Context context, Intent intent) {
        mView = view;
        mContext = context;
        mUseCase = new ArticleScreen();
        mArticleUrl = getArticleUrl(intent);
        mConnectivityCheck = new ConnectivityCheck(mContext);
    }
    
    @Override
    public void onViewCreated() {
        mView.showLoadingScreen();
    }
    
    @Override
    public void start() {
        mUseCase.getRemoteArticleDetails(mArticleUrl).subscribe(mArticleRemoteObserver);
    }
    
    @Override
    public void stop() {
        mDisposable.clear();
    }
    
    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        int itemThatWasClickedId = item.getItemId();
        if (itemThatWasClickedId == R.id.about_app) {
            DialogBuilder builder = new DialogBuilder();
            builder.showAboutDialog(mContext);
            return true;
        }
        return false;
    }
    
    @Override
    public void onRetryClick() {
        if (mConnectivityCheck.isNetworkAvailable()) {
            mUseCase.getRemoteArticleDetails(mArticleUrl).subscribe(mArticleRemoteObserver);
            mView.showLoadingScreen();
        } else {
            mView.showNetworkToast();
        }
    }
    
    private String getArticleUrl(Intent intent) {
        if (intent.hasExtra(ARTICLE_URL)) {
            return intent.getStringExtra(ARTICLE_URL);
        } else {
            mView.showErrorMessage();
            return "";
        }
    }
    
    private class ArticleRemoteObserver implements SingleObserver<Article> {
        
        @Override
        public void onSubscribe(@NonNull final Disposable d) {
            mDisposable.add(d);
        }
        
        @Override
        public void onError(Throwable e) {
            mView.showErrorMessage();
            e.printStackTrace();
        }
        
        @Override
        public void onSuccess(@NonNull final Article article) {
            mView.setImgUrl(article.getImgUrl());
            mView.setTitle(article.getTitle());
            mView.setDescritpion(article.getDescription());
            mView.showArticleDetails();
        }
    }
}

File: app/src/main/java/pl/orbitemobile/wspolnoty/activities/article/ArticlePresenter.kt
/*
 * Copyright (c) 2017. All Rights Reserved. Michal Jankowski orbitemobile.pl
 */
package pl.orbitemobile.wspolnoty.activities.article

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.PersistableBundle
import android.view.MenuItem
import pl.orbitemobile.wspolnoty.R
import pl.orbitemobile.wspolnoty.activities.article.domain.ArticleScreen
import pl.orbitemobile.wspolnoty.data.dto.ArticleDTO
import pl.orbitemobile.wspolnoty.utilities.DialogBuilder
import pl.orbitemobile.wspolnoty.utilities.isNetwork

class ArticlePresenter : ArticleContract.Presenter() {

    val mUseCase: ArticleScreen = ArticleScreen() //todo: add composite disposable to both presenters here and in mvp package
    val ARTICLE_URL = "ARTICLE_URL"

    var mArticleUrl: String? = null

    override fun onViewAttached() {
        mView?.showLoadingScreen()
        setArticleUrl()
        issueDataDownloadIfNetworkIsAvailable()
    }

    private fun setArticleUrl() {
        val intent = mView?.getIntent()
        if (intent?.hasArticleUrl() == true) {
            intent.setArticleUrl()
        }
    }

    override fun onStart(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.about_app) {
            DialogBuilder().showAboutDialog(mView?.context)
            return true
        }
        return false
    }

    override fun onRetryClick() = issueDataDownloadIfNetworkIsAvailable()

    override fun onShowWebsiteClick() {
        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(mArticleUrl))
        mView?.getContext()?.startActivity(browserIntent)
    }

    private fun issueDataDownloadIfNetworkIsAvailable() {
        if (isNetwork()) {
            mView?.showLoadingScreen()
            issueDataDownload()
        } else {
            mView?.showErrorMessage()
            mView?.showNetworkToast()
        }
    }

    private fun issueDataDownload() {
        val articleUrl = mArticleUrl
        if (articleUrl != null) {
            mUseCase.getRemoteArticleDetails(articleUrl).subscribe(this)
        } else {
            // todo: article url is null
        }
    }

    private fun Intent.hasArticleUrl(): Boolean = hasExtra(ARTICLE_URL)

    private fun Intent.setArticleUrl() {
        mArticleUrl = getStringExtra(ARTICLE_URL)
    }

    override fun onError(e: Throwable) {
        mView?.showErrorMessage() //todo: does this still work with rxUtils?
        e.printStackTrace()
    }

    override fun onSuccess(article: ArticleDTO) {
        mView?.showArticleDetails(article)
    }

}
